pkg/scuzz/ata.go

```diff
@@ -0,0 +1,281 @@
+	writePIOExt:          {},
+	writeDMAExt:          {},
+	writeFPDMA:           {},
+	readVerify:           {},
+	readVerifyOnce:       {},
+	readVerifyExt:        {},
+	writeUNCExt:          {},
+	formatTrack:          {},
+	downloadMicrocode:    {},
+	standbynow2:          {},
+	checkpowermode2:      {},
+	sleepnow2:            {},
+	pidentify:            {},
+	readNativeMax:        {},
+	readNativeMaxExt:     {},
+	smart:                {},
+	dco:                  {},
+	eraseSectors:         {},
+	readDMA:              {},
+	writeDMA:             {},
+	doorlock:             {},
+	doorunlock:           {},
+	standbynow1:          {},
+	idleimmediate:        {},
+	setidle:              {},
+	setMax:               {},
+	setMaxExt:            {},
+	setMultiple:          {},
+	checkpowermode1:      {},
+	sleepnow1:            {},
+	flushcache:           {},
+	flushcacheExt:        {},
+	identify:             {},
+	setfeatures:          {},
+	securitySetPass:      {},
+	SecurityUnlock:       {cdb: cdb{ata16, lba48 | pioOut}},
+	securityErasePrepare: {},
+	securityEraseUnit:    {},
+	securityFreezeLock:   {},
+	securityDisable:      {},
+	vendorSpecific0x80:   {},
+}
+
+const (
+	ataUsingLBA uint8 = (1 << 6)
+	ataStatDRQ        = (1 << 3)
+	ataStatErr        = (1 << 0)
+)
+
+// /*
+//  * Useful parameters for init_hdio_taskfile():
+//  */
+// enum {	RW_READ			= 0,
+// 	RW_WRITE		= 1,
+// 	LBA28_OK		= 0,
+// 	LBA48_FORCE		= 1,
+// };
+
+// enum {
+// 	/*
+// 	 * These (redundantly) specify the category of the request
+// 	 */
+// 	TASKFILE_CMD_REQ_NODATA	= 0,	/* ide: IDE_DRIVE_TASK_NO_DATA */
+// 	TASKFILE_CMD_REQ_IN	= 2,	/* ide: IDE_DRIVE_TASK_IN */
+// 	TASKFILE_CMD_REQ_OUT	= 3,	/* ide: IDE_DRIVE_TASK_OUT */
+// 	TASKFILE_CMD_REQ_RAW_OUT= 4,	/* ide: IDE_DRIVE_TASK_RAW_WRITE */
+// 	/*
+// 	 * These specify the method of transfer (pio, dma, multi, ..)
+// 	 */
+// 	TASKFILE_DPHASE_NONE	= 0,	/* ide: TASKFILE_IN */
+// 	TASKFILE_DPHASEPIO_IN	= 1,	/* ide: TASKFILE_IN */
+// 	TASKFILE_DPHASE_PIO_OUT	= 4,	/* ide: TASKFILE_OUT */
+// };
+
+const (
+	read  uint8 = 0
+	write       uint8 = 1
+
+	// ioPIO indicates we are doing programmed IO
+	ioPIO = 0
+	// ioDMA indicates we are doing DMA
+	ioDMA = 1
+
+	opCheckCondition = 0x02
+	opDriverSense    = 0x08
+
+	ata16    = 0x85
+	ata16Len = 16
+	cdbSize  = ata16Len
+
+	maxSBLen = 32
+
+	lba48    = 1
+	nonData  = (3 << 1)
+	pioIn    = (4 << 1)
+	pioOut   = (5 << 1)
+	protoDMA = (6 << 1)
+)
+
+const (
+	tlenNoData = 0 << 0
+	tlenFeat   = 1 << 0
+	tlenNsect  = 2 << 0
+
+	tlenBytes   = 0 << 2
+	tlenSectors = 1 << 2
+
+	tdirTo    = 0 << 3
+	tdirDrom  = 1 << 3
+	checkCond = 1 << 5
+)
+
+// the cdb struct is used to create the cdb array. We do not
+// export this struct as it is unlikely most users would use it
+// correctly.
+type commandDataBlock struct {
+	cmd   Cmd
+	lba   uint64
+	nsect uint16
+	dev   uint8
+	dma   bool
+	write bool
+	data  []byte
+}
+
+type cdb [cdbSize]byte
+type sb [maxSBLen]byte
```

Comment:
what's an sb

Reply:
spell it out

Reply:
to be honest, I 'm still not sure. It might be a status block. I have not found a use for it yet. It's just all over the linux sg layer.